### PR TITLE
Bugfix: ansiblegate, issue 44670

### PR DESCRIPTION
**Pull request by isbm**
### What does this PR do?

Blocker/P1 bugfix.

### What issues does this PR fix or reference?

Fixes blocker, described at https://github.com/saltstack/salt/issues/44670

### Tests written?

Yes

### CODE REVIEW

File: salt/modules/ansiblegate.py [hunk 1]
```diff
@@ -215,13 +215,13 @@ def __virtual__():
     ret = ansible is not None
     msg = not ret and "Ansible is not installed on this system" or None
     if msg:
-        log.warning(msg)
+        log.debug(msg)
```
Comment (rallytime) [inline salt/modules/ansiblegate.py]: Actually, we shouldn't be logging in `__virtual__()` at all. In the past, this was not well-documented, but I just added some docs for this the other day in PR #44588. We either need to not log in here at all, or this needs to be at the `trace` level.

**Comment by rallytime** [comment]
Thanks for addressing this so quickly @isbm. I just have one comment here.

**Comment by isbm** [comment]
> Thanks for addressing this so quickly @isbm. I just have one comment here.

@rallytime you're working with German SUSE, what else you would expect. :sunglasses: Done!

File: salt/modules/ansiblegate.py [hunk 2]
```diff
@@ -215,13 +215,13 @@ def __virtual__():
     ret = ansible is not None
     msg = not ret and "Ansible is not installed on this system" or None
     if msg:
-        log.warning(msg)
+        log.trace(msg)
```
Comment (mirceaulinic) [inline salt/modules/ansiblegate.py]: Please remove.

**Review by mirceaulinic — CHANGES_REQUESTED** [review]
I'd suggest using:

```python
 try:
     import ansible
     import ansible.constants
     import ansible.modules
     HAS_ANSIBLE = True
 except ImportError:
     HAS_ANSIBLE = False
```

rather than:

```python
 try:
     import ansible
     import ansible.constants
     import ansible.modules
 except ImportError:
     ansible = None
```

This is the way the execution modules usually are written (without this being a hard rule, but I find it nicer, more effective and also reduces the complexity of the code), e.g., https://github.com/saltstack/salt/pull/44764/commits/9807acf1bfa4de8998a7d6509bfeb666f19bf174 (feel free to cherry-pick the commit if you wish). Additionally, that's the way it is recommended: https://docs.saltstack.com/en/latest/ref/modules/#returning-error-information-from-virtual.

Also, please remove the log line from `__virtual__`: https://docs.saltstack.com/en/latest/ref/modules/#logging-restrictions

> As a rule, logging should not be done anywhere in a Salt module before it is loaded. This rule apples to all code that would run before the __virtual__() function, as well as the code within the __virtual__() function itself.

**Comment by isbm** [comment]
@mirceaulinic the way I see it that `HAS_FOO` is like tracking that some `foo` is `True` or not. I prefer minimalism as possible, so I would not create a sort of reference to the data instead to just ask the data. I find this as same as anytime we do:

```python
try:
    something = operation()
except OperationFailure:
    something = None

if something is not None:
    ...
```

You don't do:
```python
try:
    something = operation()
    SOMETHING_WENT_RIGHT = True
except OperationFailure:
    something = None
    SOMETHING_WENT_RIGHT = False

if SOMETHING_WENT_RIGHT:
    something ...
    ...
```

So I see no reason why that should be different during the module import, since the `ansible` is just another variable, with type `<module>`, once it is imported. The same way it can happily be just `None` and that very type can be just checked. I would understand if `import ansible` would be an addition to other things, then it is "easier" (still debatable, BTW) to read the source. But if I have to shut down `__virtual__` just immediately, then I see no reason allocate more memory just in oder to bail out, because `ansible` is _anyway_ set to `None`.

As of logging, I put that to `trace` as per @rallytime .

**Comment by mirceaulinic** [comment]
Sure, I understand you personal preference, but that's the way it is both recommended, and Salt developers prefer (as you can see going through the existing modules): https://docs.saltstack.com/en/latest/ref/modules/#returning-error-information-from-virtual

**Comment by isbm** [comment]
@mirceaulinic I don't think I am going to discuss that any further. We have quite serious issues right now with systemd on SLES and RHELs supporting Salt, and thus additional logging at trace level would be actually way more helpful than what it is right now. Unless @rallytime is going to rule out differently, it stays. Sorry.

**Comment by mirceaulinic** [comment]
No problem, I was trying to be helpful. For your info, when the `__virtual__` returns `False, 'Ansible is not installed on this system'`, the message 'Ansible is not installed on this system' will be logged at the debug level. Thanks for your addition @isbm! :)

**Comment by isbm** [comment]
@mirceaulinic https://github.com/saltstack/salt/pull/44691/commits/79d27d6a307b178a9de5a1506bcee483a4bad797 — hopefully this makes you happy. :wink:

**Review by rallytime — APPROVED** [review]

**Comment by rallytime** [comment]
re-run lint

**Pull request merged**